File: app/src/main/java/com/appharbr/example/app/gam/GamCombinationNativeAdWithBannerActivity.java
package com.appharbr.example.app.gam;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.appharbr.example.app.R;
import com.appharbr.sdk.engine.AdResult;
import com.appharbr.sdk.engine.AdSdk;
import com.appharbr.sdk.engine.AdStateResult;
import com.appharbr.sdk.engine.AppHarbr;
import com.appharbr.sdk.engine.listeners.AHListener;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.VideoController;
import com.google.android.gms.ads.admanager.AdManagerAdRequest;
import com.google.android.gms.ads.admanager.AdManagerAdView;
import com.google.android.gms.ads.formats.AdManagerAdViewOptions;
import com.google.android.gms.ads.nativead.NativeAd;
import com.google.android.gms.ads.nativead.NativeAdView;

import java.util.Arrays;
import java.util.Locale;

public class GamCombinationNativeAdWithBannerActivity extends AppCompatActivity {

    private Button refresh;
    private TextView videoStatus;
    private AdManagerAdView currentBannerAdView;
    private NativeAd unifiedNativeAd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.gam_nativead_activity);

        refresh = findViewById(R.id.btn_refresh);
        videoStatus = findViewById(R.id.tv_video_status);
        refresh.setOnClickListener(v -> requestAdLoader());

        requestAdLoader();
    }

    private void requestAdLoader() {
        refresh.setEnabled(false);
        AdLoader adLoader = new AdLoader.Builder(getApplicationContext(),
                getApplicationContext().getResources().getString(R.string.gam_nativebanner_ad_unit_id))
                .forNativeAd(nativeAd -> {
                    releaseBannerResources();

                    // ############ Check with AppHarbr if this ad suppose to be display ################
                    AdResult adResult = AppHarbr.shouldBlockNativeAd(AdSdk.GAM, nativeAd, getApplicationContext().getResources().getString(R.string.gam_nativebanner_ad_unit_id));
                    if (adResult.getAdStateResult() == AdStateResult.BLOCKED) {
                        // **** AppHarbr Blocked The Native Ad - Do Not Render it. ****
                        // May request another ad.
                        refresh.setEnabled(true);
                        return;
                    }

                    if (unifiedNativeAd != null) {
                        unifiedNativeAd.destroy();
                    }
                    unifiedNativeAd = nativeAd;

                    FrameLayout frameLayout = findViewById(R.id.fl_adplaceholder);
                    NativeAdView adView = (NativeAdView) getLayoutInflater().inflate(R.layout.ad_unified, null);
                    populateUnifiedNativeAdView(nativeAd, adView);
                    frameLayout.removeAllViews();
                    frameLayout.addView(adView);
                })
                .forAdManagerAdView(adView -> {
                    refresh.setEnabled(true);
                    releaseBannerResources();
                    currentBannerAdView = adView;

                    // ####### Publisher got a new Banner - Using AppHarbr to monitor it ########
                    AppHarbr.addBannerViewFromAdLoader(AdSdk.GAM, adView, getLifecycle(), ahListener);

                    FrameLayout frameLayout =
                            findViewById(R.id.fl_adplaceholder);
                    frameLayout.removeAllViews();
                    frameLayout.addView(adView);
                }, new AdSize(300, 250)).withAdListener(new AdListener() {
                    @Override
                    public void onAdFailedToLoad(LoadAdError error) {
                    }
                })
                .withAdManagerAdViewOptions(new AdManagerAdViewOptions.Builder()
                        .build())
                .build();
        adLoader.loadAd(new AdManagerAdRequest.Builder().build());
    }

    // Publisher to display Native Ad
    private void populateUnifiedNativeAdView(NativeAd nativeAd, NativeAdView adView) {
        adView.setMediaView(adView.findViewById(R.id.ad_media));
        adView.setHeadlineView(adView.findViewById(R.id.ad_headline));
        adView.setBodyView(adView.findViewById(R.id.ad_body));
        adView.setCallToActionView(adView.findViewById(R.id.ad_call_to_action));
        adView.setIconView(adView.findViewById(R.id.ad_app_icon));
        adView.setPriceView(adView.findViewById(R.id.ad_price));
        adView.setStarRatingView(adView.findViewById(R.id.ad_stars));
        adView.setStoreView(adView.findViewById(R.id.ad_store));
        adView.setAdvertiserView(adView.findViewById(R.id.ad_advertiser));
        if (nativeAd.getBody() == null) {
            adView.getBodyView().setVisibility(View.INVISIBLE);
        } else {
            adView.getBodyView().setVisibility(View.VISIBLE);
            ((TextView) adView.getBodyView()).setText(nativeAd.getBody());
        }
        if (nativeAd.getCallToAction() == null) {
            adView.getCallToActionView().setVisibility(View.INVISIBLE);
        } else {
            adView.getCallToActionView().setVisibility(View.VISIBLE);
            ((Button) adView.getCallToActionView()).setText(nativeAd.getCallToAction());
        }
        if (nativeAd.getIcon() == null) {
            adView.getIconView().setVisibility(View.GONE);
        } else {
            ((ImageView) adView.getIconView()).setImageDrawable(
                    nativeAd.getIcon().getDrawable());
            adView.getIconView().setVisibility(View.VISIBLE);
        }
        if (nativeAd.getPrice() == null) {
            adView.getPriceView().setVisibility(View.INVISIBLE);
        } else {
            adView.getPriceView().setVisibility(View.VISIBLE);
            ((TextView) adView.getPriceView()).setText(nativeAd.getPrice());
        }
        if (nativeAd.getStore() == null) {
            adView.getStoreView().setVisibility(View.INVISIBLE);
        } else {
            adView.getStoreView().setVisibility(View.VISIBLE);
            ((TextView) adView.getStoreView()).setText(nativeAd.getStore());
        }
        if (nativeAd.getStarRating() == null) {
            adView.getStarRatingView().setVisibility(View.INVISIBLE);
        } else {
            ((RatingBar) adView.getStarRatingView())
                    .setRating(nativeAd.getStarRating().floatValue());
            adView.getStarRatingView().setVisibility(View.VISIBLE);
        }
        if (nativeAd.getAdvertiser() == null) {
            adView.getAdvertiserView().setVisibility(View.INVISIBLE);
        } else {
            ((TextView) adView.getAdvertiserView()).setText(nativeAd.getAdvertiser());
            adView.getAdvertiserView().setVisibility(View.VISIBLE);
        }
        adView.setNativeAd(nativeAd);
        VideoController vc = nativeAd.getMediaContent().getVideoController();
        if (vc.hasVideoContent()) {
            videoStatus.setText(String.format(Locale.getDefault(),
                    "Video status: Ad contains a video asset."));
            vc.setVideoLifecycleCallbacks(new VideoController.VideoLifecycleCallbacks() {
                @Override
                public void onVideoEnd() {
                    refresh.setEnabled(true);
                    videoStatus.setText("Video status: Video playback has ended.");
                    super.onVideoEnd();
                }
            });
        } else {
            videoStatus.setText("Video status: Ad does not contain a video asset.");
            refresh.setEnabled(true);
        }
    }


    private final AHListener ahListener = (view, unitId, adFormat, reasons)
            -> Log.d("LOG", "AppHarbr - onAdBlocked for: " + unitId + ", reason: " + Arrays.toString(reasons));

    private void releaseBannerResources() {
        if (currentBannerAdView != null) {
            AppHarbr.removeBannerView(currentBannerAdView);
            currentBannerAdView.destroy();
            currentBannerAdView = null;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        releaseBannerResources();
    }
}

File: kotlinsampleapp/src/main/java/com/appharbr/kotlin/example/app/gam/GamCombinationNativeAdWithBannerActivity.kt
package com.appharbr.kotlin.example.app.gam

import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.core.graphics.drawable.toBitmap
import com.appharbr.kotlin.example.app.R
import com.appharbr.kotlin.example.app.ui.theme.AppHarbrExampleAppTheme
import com.appharbr.sdk.engine.AdBlockReason
import com.appharbr.sdk.engine.AdSdk
import com.appharbr.sdk.engine.AdStateResult
import com.appharbr.sdk.engine.AppHarbr
import com.appharbr.sdk.engine.adformat.AdFormat
import com.appharbr.sdk.engine.listeners.AHListener
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.admanager.AdManagerAdRequest
import com.google.android.gms.ads.admanager.AdManagerAdView
import com.google.android.gms.ads.formats.AdManagerAdViewOptions
import com.google.android.gms.ads.nativead.NativeAd
import java.util.*

class GamCombinationNativeAdWithBannerActivity : ComponentActivity() {

    private val nativeAdState = mutableStateOf<NativeAd?>(null)
    private val adManagerAdViewState = mutableStateOf<AdManagerAdView?>(null)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestCombinedAd()

        setContent {
            AppHarbrExampleAppTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colors.background
                ) {

                    Column(
                        modifier = Modifier.fillMaxSize(),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        DisplayNativeAd()
                        DisplayBanner()
                    }
                }
            }
        }
    }

    private fun requestCombinedAd() {
        AdLoader.Builder(
            applicationContext,
            applicationContext.resources.getString(R.string.gam_nativebanner_ad_unit_id)
        )
            .forNativeAd { nativeAd: NativeAd ->

                // ############ Check with AppHarbr if this ad suppose to be display ################
                val adResult = AppHarbr.shouldBlockNativeAd(AdSdk.GAM, nativeAd, applicationContext.resources.getString(R.string.gam_nativebanner_ad_unit_id))
                if (adResult.adStateResult == AdStateResult.BLOCKED) {
                    // **** AppHarbr Blocked The Native Ad - Do Not Render it. ****
                    // May request another ad.
                    nativeAdState.value = null
                }

                if (isDestroyed) {
                    nativeAdState.value?.destroy()
                    return@forNativeAd
                }
                nativeAdState.value?.destroy()
                nativeAdState.value = nativeAd
            }
            .forAdManagerAdView({ adView: AdManagerAdView ->

                // ####### Publisher got a new Banner - Using AppHarbr to monitor it ########
                adManagerAdViewState.value?.let {
                    AppHarbr.removeBannerView(it)
                }
                AppHarbr.addBannerViewFromAdLoader(AdSdk.GAM, adView, ahListener)

                adManagerAdViewState.value = adView

            }, AdSize(300, 250)).withAdListener(object : AdListener() {
                override fun onAdFailedToLoad(error: LoadAdError) {}
            })
            .withAdManagerAdViewOptions(AdManagerAdViewOptions.Builder().build()).build()
            .loadAd(AdManagerAdRequest.Builder().build())
    }

    @Composable
    private fun DisplayNativeAd() {
        nativeAdState.value?.let {
            Text(text = "Headline: ${it.headline ?: "Empty Headline"}")
            Text(text = "Body: ${it.body ?: "Empty body"}")
            it.icon?.drawable?.toBitmap()?.let { bitmap -> Image(bitmap.asImageBitmap(), "") }
            for (image in it.images) {
                image.drawable?.toBitmap()?.let { bitmap -> Image(bitmap.asImageBitmap(), "") }
            }
        } ?: kotlin.run {
            Text(
                text = stringResource(id = R.string.gam_native_and_banner),
                fontSize = 20.sp
            )
            CircularProgressIndicator()
        }
    }

    @Composable
    private fun DisplayBanner() {
        adManagerAdViewState.value?.let {
            AndroidView(
                modifier = Modifier.wrapContentSize(),
                factory = { _ ->
                    it
                })
        }
    }

    private val ahListener =
        AHListener { view: Any?, unitId: String?, adFormat: AdFormat?, reasons: Array<AdBlockReason?>? ->
            Log.d(
                "LOG",
                "AppHarbr - onAdBlocked for: $unitId, reason: " + Arrays.toString(
                    reasons
                )
            )
        }

}